internal/kms/kmip.go

```diff
@@ -0,0 +1,527 @@
+
+	// kmipDefaultWriteTimeout is the default write network timeout.
+	kmipDefaultWriteTimeout = 10
+
+	// KMIP version.
+	protocolMajor = 1
+	protocolMinor = 4
+
+	// nonceSize is required to generate nonce for encrypting DEK.
+	nonceSize = 16
+
+	// kmipDefaultSecretsName is the default name of the Kubernetes Secret
+	// that contains the credentials to access the KMIP server. The name of
+	// the Secret can be configured by setting the `KMIP_SECRET_NAME`
+	// option.
+	//
+	// #nosec:G101, value not credential, just references token.
+	kmipDefaultSecretsName = "ceph-csi-kmip-credentials"
+
+	kmipEndpoint      = "KMIP_ENDPOINT"
+	kmipTLSServerName = "TLS_SERVER_NAME"
+	kmipReadTimeOut   = "READ_TIMEOUT"
+	kmipWriteTimeOut  = "WRITE_TIMEOUT"
+
+	// The following options are part of the Kubernetes Secrets.
+	//
+	// #nosec:G101, value not credential, just configuration keys.
+	kmipSecretNameKey    = "KMIP_SECRET_NAME"
+	kmipCACert           = "CA_CERT"
+	kmipCLientCert       = "CLIENT_CERT"
+	kmipClientKey        = "CLIENT_KEY"
+	kmipUniqueIdentifier = "UNIQUE_IDENTIFIER"
+)
+
+var _ = RegisterProvider(Provider{
+	UniqueID:    kmsTypeKMIP,
+	Initializer: initKMIPKMS,
+})
+
+type kmipKMS struct {
+	// basic options to get the secret
+	secretName string
+	namespace  string
+
+	// standard KMIP configuration options
+	endpoint         string
+	tlsConfig        *tls.Config
+	uniqueIdentifier string
+	readTimeout      uint8
+	writeTimeout     uint8
+}
+
+func initKMIPKMS(args ProviderInitArgs) (EncryptionKMS, error) {
+	kms := &kmipKMS{
+		namespace: args.Namespace,
+	}
+
+	// get secret name if set, else use default.
+	err := setConfigString(&kms.secretName, args.Config, kmipSecretNameKey)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	} else if errors.Is(err, errConfigOptionMissing) {
+		kms.secretName = kmipDefaultSecretsName
+	}
+
+	err = setConfigString(&kms.endpoint, args.Config, kmipEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// optional
+	serverName := ""
+	err = setConfigString(&serverName, args.Config, kmipTLSServerName)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+
+	// optional
+	timeout := kmipDefaulfReadTimeout
+	err = setConfigInt(&timeout, args.Config, kmipReadTimeOut)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+	kms.readTimeout = uint8(timeout)
+
+	// optional
+	timeout = kmipDefaultWriteTimeout
+	err = setConfigInt(&timeout, args.Config, kmipWriteTimeOut)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+	kms.writeTimeout = uint8(timeout)
+
+	// read the Kubernetes Secret with CA cert, client cert, client key
+	// & key unique identifier.
+	secrets, err := kms.getSecrets()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secrets: %w", err)
+	}
+
+	caCert, found := secrets[kmipCACert]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, kmipCACert)
+	}
+
+	clientCert, found := secrets[kmipCLientCert]
```

Comment:
As this is required, a client can only connect to the KMIP server with its own client certificate? That is not very common for TLS connections. Where does the KMIP standard describe this requirement?

Reply:
https://thalesdocs.com/ctp/cm/latest/reference/kmip-ref/

There are a couple of options of authenticating with kmip server, currently we are going this route. 
We can adapt this based later as required, since this just preliminary exploration with kmip encryption type.

Reply:
OK, so this is a common process for Thales CyberTrust. Have you considered adding an option like `"AUTH_METHOD": "TLSCERT"` or something? That suggests making it more modular in the future. Of course, this can be the default if no `AUTH_METHOD` is set too...

Reply:
https://github.com/ceph/ceph-csi/issues/3316
noted :+1: